protocol/messenger_backup_handler.go

```diff
@@ -0,0 +1,205 @@
+package protocol
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/status-im/status-go/images"
+	"github.com/status-im/status-go/multiaccounts/settings"
+	"github.com/status-im/status-go/protocol/protobuf"
+	"github.com/status-im/status-go/protocol/wakusync"
+)
+
+const (
+	// timeToPostponeBackedUpMessagesHandling - the idea is to wait for 30 secs in loading state with the hope that at least
+	// one message from the set of last backed up messages will arrive, that we can know which set of messages to work with
+	timeToPostponeBackedUpMessagesHandling = 30 * time.Second
```

Comment:
How do you know which one is the right one? Does this run everytime the user login? i.e we always delay processing of back ups even if it's not the first login?

I don't quite understand how we handle multiple backups (from different devices for example) or older backups from the same device, is that something that the code takes into consideration?

Reply:
An idea is to wait for 30 seconds and proceed with applying most recent **known** backup we receive within the timeframe of 30 seconds. That means newer backup messages may arrive later and they will be applied later, but older backup messages than those known as the newest during that time frame will be applied firstly.

Yes, aggregating backup messages is happening for each login for the period of 30 seconds.
We don't need to know which device a backup is coming from if it's the most recent one since all messages backed up  in a single go have the same time.

Example If device 1 backed up its state at time t1, marking that time as D1T1, now following that logic if we have backups done in this order `D1T1` < `D2T1` < `D1T2` < `D3T1` and let's say that for each of these backups 5 messages were backed up.

Now when user is recovering an account at time `T` which is now (means `D3T1` < `T`) some or all of messages backed up from any devices `D1`, `D2`, `D3` may or may not arrive within the time of 30 seconds after app start, but how we handle them is:
- timer of 30 secs started
- first arrives msg `#4` from `D2T1` we add it to queue but don't process
- then arrives msg `#2` from `D2T1` we add it to queue but don't process 
- then arrives msg `#3` from `D1T1` since it has less time that stored messages we just skip it
- then arrives msg `#4` from `D1T1` we skip it
- then arrives msg `#1` from `D1T2` we remove all messages from queue since they have less time than the most recent one and add that one
- then arrives msg `#2` from `D1T1` we skip it
- then arrives msg `#3` from `D1T2` we add it
- and now time's up, 30 seconds passed

what do we do now, we process messages from the queue even 0 messages from `D3T1` arrived, when other messages arrives later, they will be processed immediately and data they are carrying will be applied if the time for that data is greater then the time of last change.

Reply:
I see, thanks for the explanation.

One thing to note is that you can't discard backups from different devices, since they might not have converged yet if the user hasn't paired, and if you do that the risk is that they will not. 
You can safely discard though backups from the same device with different timestamps, keeping the one with the oldest timestamp as you are already doing.

Regarding the 30s, I don't think that's acceptable to delay processing, as it would impact every login and that's just not worth it, moreover is a purely client side concern and can be addressed in the client if the client wishes to do so (I still think the feature is just not worth the effort and we should just show a loading screen, this is not a solvable problem).

It's also not really addressing completely the issue since it might be that connecting to a store node is slow for example, the user might be offline, etc. Solutions like this were tried in the past and never really worked, so I don't think it's worth the complexity.

Reply:
@cammellos I applied all other suggestions you gave me, you can check them in [this commit](https://github.com/status-im/status-go/pull/2976/commits/4ed705ae2d59c51a8a379bd5f6d1c2e9b865a6c3).

As of this comment I am not sure, do I need to do something? Or we can leave that for later?
It was John's idea to wait for 30 secs and proceed with the messages which are know at that moment (that also doesn't stop us from applying other messages if they appear later) and to do that on the `stauts-go` side cause the same logic should be for both mobile and desktop apps.